numpy/core/tests/test_multiarray.py

```diff
@@ -7938,3 +7938,48 @@ def test_getfield():
     pytest.raises(ValueError, a.getfield, 'uint8', -1)
     pytest.raises(ValueError, a.getfield, 'uint8', 16)
     pytest.raises(ValueError, a.getfield, 'uint64', 0)
+
+def test_multiarray_module():
+    # gh-12736
+    # numpy 1.16 replaced the multiarray and umath c-extension modules with
+    # a single _multiarray_umath one. For backward compatibility, it added a
+    # pure-python multiarray.py and umath.py shim so people can still do
+    # from numpy.core.multirarray import something-public-api
+    # It turns out pip can leave old pieces of previous versions of numpy
+    # around when installing a newer version. If the old c-extension modules
+    # are found, they will be given precedence to the new pure-python ones.
```

Comment:
Is this behavior documented anywhere? This seems like a design flaw in CPython...

Reply:
A preference for c-extension modules over pure-python ones is baked into cpython. [`_get_supported_file_loaders`](https://github.com/python/cpython/blob/v3.7.1/Lib/importlib/_bootstrap_external.py#L1482) seems to be the preference ranking in python3. In python 2 there is [this](https://github.com/python/cpython/blob/2.7/Lib/imputil.py#L78) explicit statement of preference 

> "The default will import dynamic-load modules first, followed by .py files (or a .py file's cached bytecode)"